### DISCUSSION TITLE
use RDT for get_orders

### DISCUSSION THREAD

**Discussion by LuisB2904** [saleweaver/python-amazon-sp-api#256] [Q&A] [answered]
i want to get the buyerInfo from the get_orders call. I've found this code in another discussion and it works fine for me so i get a RDT. But when i want to call get_orders with my RDT i get this error

`sp_api.base.exceptions.SellingApiServerException: [{'message': 'We encountered an internal error. Please try again.', 'code': 'InternalFailure'}]`

My Code:
```
credentials = dict(
    refresh_token=xxx,
    lwa_client_secret=xxx,
    aws_secret_key=xxx,
    aws_access_key=xxx,
    role_arn=xxx
)



class MyOrders(Orders):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)

    @sp_endpoint('/orders/v0/orders')
    def get_orders(self, **kwargs) -> ApiResponse:
        headers = kwargs.pop('headers') if 'headers' in kwargs else None
        return self._request(kwargs.pop('path'), params={**kwargs}, headers=headers)

tokens_api = Tokens(credentials=credentials)
response = tokens_api.create_restricted_data_token(
    restrictedResources=[
        {
            "method": "GET",
            "path": "/orders/v0/orders",
            "dataElements": ["buyerInfo", "shippingAddress"]
        }
    ]
)
restrictedDataToken = response.payload['restrictedDataToken']
print(restrictedDataToken)
orders_api = MyOrders(credentials=credentials)

response = orders_api.get_orders(
    #CreatedAfter=(datetime.utcnow() - timedelta(days=2)).isoformat(),
    headers={'x-amz-access-token':restrictedDataToken}

)
```

Can you help me?

**Comment by saleweaver** [comment]
I haven't seen this error outside of sandbox mode yet, are you sure this is a live request?

**Comment by saleweaver** [comment, marked as answer]
Hi, I've updated the package. Please check out https://github.com/saleweaver/python-amazon-sp-api/releases/tag/v0.7.0. You can pass the token to the client like this:

```
Orders(restricted_data_token='......').get_orders(...)
```

**Reply by YamashitaEdit** [reply]
@saleweaver 

Hi, I need some help with this update.
Do you know why an error "[{'message': 'Access to requested resource is denied.', 'code': 'Unauthorized', 'details': ''}]" happens with my code bellow?

I want to access PII data and get buyer info.
Now I can get buyer info properly, but when I comment out `# restricted_data_token=restrictedDataToken`, this code doesn't work and the error will occur.

```                tokens_api = Tokens(credentials=credentials, marketplace=Marketplaces.US)
                response = tokens_api.create_restricted_data_token(restrictedResources=body)
                restrictedDataToken = response.payload['restrictedDataToken']
                response = Orders(
                    credentials=credentials,
                    marketplace=Marketplaces.US,
                    # restricted_data_token=restrictedDataToken
                ).get_orders(
                    MarketplaceIds='ATVPDKIKX0DER',
                    AmazonOrderIds=OrderID
                )
```

Thank you in advance.

**Reply by saleweaver** [reply]
- You you get the rdt at all? 

- Was the pii access permission added recently to the app?

- Did you get a new refresh token yet? 

Please post error output. Thx.

**Reply by YamashitaEdit** [reply]
@saleweaver 

・You you get the rdt at all?
Yes

・Was the pii access permission added recently to the app?
I just have added the permission a few hours ago.
I have an restricted data token returned with `restrictedDataToken = response.payload['restrictedDataToken']` properly.

・Did you get a new refresh token yet?
After adding the pii access permission, I didn't get it yet?
Need to refresh it?

Thank you.

**Reply by saleweaver** [reply]
Most likely it's just that you need a new one. You have to reauthorize the app whenever permissions have changed. 
Make sure that you app has permissions for orders as well.

**Reply by saleweaver** [reply]
There's a shorthand in the api as well. To receive orders with pii data, you can simply use

```
orders = Orders().get_orders(
    RestrictedResources=['buyerInfo', 'shippingAddress'],
    LastUpdatedAfter=(datetime.utcnow() - timedelta(days=1)).isoformat()
)
```

**Reply by YamashitaEdit** [reply]
I just tried this and I got an error like "[{'code': 'InvalidInput', 'message': 'Application does not have access to one or more requested data elements: [buyerInfo, shippingAddress]', 'details': ''}]"...

```
orders = Orders(credentials=credentials, marketplace=Marketplaces.US).get_orders(
    RestrictedResources=['buyerInfo', 'shippingAddress'],
    LastUpdatedAfter=(datetime.utcnow() - timedelta(days=1)).isoformat()
)

**Reply by saleweaver** [reply]
and did you try to remove one of the dataelements, then the other? If you get this error, you'll have to check permissions for your app

**Reply by YamashitaEdit** [reply]
>Most likely it's just that you need a new one. You have to reauthorize the app whenever permissions have changed.
  Make sure that you app has permissions for orders as well.

Thank you I will get a new one

btw, I have checked all check-boxes
![image](https://user-images.githubusercontent.com/91242986/148094307-bdd46daf-6f8c-4d4d-bad3-1dccb6296b84.png)

**Reply by saleweaver** [reply]
Can you have a look in your developer profile and check if all permissions are assigned?

**Reply by YamashitaEdit** [reply]
developer profile of amazon seller central? or something other?
Sorry for the rudimentary question

**Reply by saleweaver** [reply]
yes in seller central. The `Your developer profile` link, right under `Developer Central`

**Reply by YamashitaEdit** [reply]
Appreciate it

Maybe I haven't got enough permissions ...?
![image](https://user-images.githubusercontent.com/91242986/148096622-5c1b007a-48a9-4e7d-b627-1243cd113481.png)

**Reply by saleweaver** [reply]
Yeah, that's the problem. You need to apply for pii access first. 

Here's some documentation on the questions, you need to follow them closely. 

https://sellercentral.amazon.com/mws/static/policy?documentType=DPP
https://sellercentral.amazon.com/mws/static/policy?documentType=AUP

**Reply by YamashitaEdit** [reply]
I completely overlooked it.
I will work on it again tomorrow and update the results here.

Thank you so much I Appreciate it all for today.

**Reply by YamashitaEdit** [reply]
I was going to register and apply for a developer profile to get PII data access privileges, but I decided against it this time because I was not in a position to meet the security management rules as a individual developer .

But all the answers were really helpful.
Thank you very much for this time.
